### PR TITLE
long read postprocess

### PR DESCRIPTION
**Pull request by morsecodist**

### CODE REVIEW

File: workflows/long-read-mngs/run.wdl [hunk 1]
```diff
@@ -716,11 +872,47 @@ workflow czid_long_read_mngs {
...
+    call ReassignM8NR {
+      input:
+        m8 = RunCallHitsNR.deduped_out_m8,
+        reads_to_contigs_tsv = RunReadsToContigs.reads_to_contigs_tsv,
+        docker_image_id = docker_image_id,
+    }
+
+    call GenerateAnnotatedFasta {
+      input:
+        pre_alignment_fasta = RunSubsampling.subsampled_fastq,
```
Comment (rzlim08) [inline workflows/long-read-mngs/run.wdl:895]: The function is supposed to take in a fasta but the output of subsampling is a fastq. we could use `https://bioinf.shenwei.me/seqkit/usage/#fq2fa` to convert

Comment (morsecodist) [reply workflows/long-read-mngs/run.wdl:1204]: Thanks!

File: lib/bin/reassign_m8.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        "read_id",
+        "contig_id",
+        "alignment",
+    ], usecols=[
+        "read_id",
+        "contig_id",
+    ])
+    reads_to_contigs = reads_to_contigs[reads_to_contigs.contig_id != "*"]
+    reassigned = pd.merge(m8, reads_to_contigs, left_on="read_or_contig_id", right_on="contig_id", how="left")
+    reassigned["read_or_contig_id"] = reassigned.apply(lambda row: row["read_id"] or row["read_or_contig_id"], axis=1)
```
Comment (rzlim08) [inline lib/bin/reassign_m8.py:36]: Use pd.isnull to cover the case that the empty row gets read in as NaN (not falsey in python)
```suggestion
    reassigned["read_or_contig_id"] = reassigned.apply(lambda row: row["read_id"] if not pd.isnull(row["read_id"]) else row["read_or_contig_id"], axis=1)
    
```

File: lib/bin/reassign_m8.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        "gap_openings",
+        "q_start",
+        "q_end",
+        "s_start",
+        "s_end",
+        "evalue",
+        "bit_score",
+    ])
+
+    reads_to_contigs = pd.read_csv(reads_to_contigs_filepath, sep="\t", names=[
```
Comment (rzlim08) [inline lib/bin/reassign_m8.py:26]: I think there can be duplicates in the reads_to_contigs tsv due to a read being able to align multiple times to a contig.

File: workflows/long-read-mngs/run.wdl [hunk 4]
```diff
@@ -211,12 +263,14 @@ task RunReadsToContigs {
 
         # convert non-contigs.fastq file to .fasta file
         seqtk seq -a sample.non_contigs.fastq > sample.non_contigs.fasta
+        cat sample.reads_to_contigs.sam | grep -v "^@" | cut -f1,3,10 > reads_to_contigs.tsv
```
Comment (rzlim08) [inline workflows/long-read-mngs/run.wdl:266]: maybe adding `--secondary=no` to minimap2 might be good to keep multiple hits from the output.

File: workflows/long-read-mngs/run.wdl [hunk 5]
```diff
@@ -575,73 +1064,91 @@ workflow czid_long_read_mngs {
         File lineage_db
         File accession2taxid_db
         File taxon_blacklist
+        File taxon_whitelist = "s3://czid-public-references/taxonomy/2020-02-10/respiratory_taxon_whitelist.txt"
```
Comment (rzlim08) [inline workflows/long-read-mngs/run.wdl:1067]: I don't think we should be using a whitelist by default

File: workflows/long-read-mngs/run.wdl (hunk 4 above)
Comment (rzlim08) [reply workflows/long-read-mngs/run.wdl:266]: actually I don't think this works

File: lib/bin/reassign_m8.py (hunk 2 above)
Comment (morsecodist) [reply lib/bin/reassign_m8.py]: Got rid of this file

File: lib/bin/reassign_m8.py (hunk 3 above)
Comment (morsecodist) [reply lib/bin/reassign_m8.py]: Got rid of this file

File: workflows/long-read-mngs/run.wdl [hunk 6]
```diff
@@ -545,12 +643,403 @@ task UnmappedReads {
...
+        python3 <<CODE
+        from idseq_dag.steps.blast_contigs import generate_contig_taxon_summary
+
+        generate_contig_taxon_summary(
+            "~{read_to_contig_tsv}",
+            "~{m8}",
+            "~{hitsummary}",
+            "nt",
+            "~{deuterostome_db}",
+            "~{taxon_whitelist}",
```
[the remark this answers is not in the document]
Comment (morsecodist) [reply workflows/long-read-mngs/run.wdl]: I do disable it if it is not provided

**Comment by morsecodist** [comment]
This PR has been extensively tested in staging so I am merging so I can release a normal version for deployment

**Pull request merged**